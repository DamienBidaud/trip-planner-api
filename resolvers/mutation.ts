import { sign } from "jsonwebtoken";
import { hash, compare } from "bcryptjs";
import { Context } from "../index.d";

const signup = async (
  _: unknown,
  args: { email: string; password: string; username: string },
  context: Context
) => {
  const password = await hash(args.password, 10);
  const author = await context.prisma.author.create({
    data: {
      ...args,
      password,
    },
  });

  const token = sign(
    {
      authorId: author.id,
    },
    "DEV_ENV"
  );

  return {
    token,
    author,
  };
};

const login = async (_: unknown, args: { email: string; password: string}, context: Context) => {
  const author = await context.prisma.author.findOne(
    {
      where: {
        email: args.email,
      },
    }
  );

  if (!author) {
    throw new Error(`No author with the email ${args.email}.`);
  }

  const valid = await compare(args.password, author.password);

  if (!valid) {
    throw new Error(`The password is incorrect.`);
  }

  const token = sign(
    {
      authorId: author.id,
    },
    "DEV_ENV"
  );

  return {
    token,
    author,
  };
};

export default { signup, login };
