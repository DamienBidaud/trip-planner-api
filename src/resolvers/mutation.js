const bcrypt = require("bcryptjs");
const jwt = require("jsonwebtoken");
const { findLocation, addLocation, getActivity, getTrip } = require("../utils");

const signup = async (root, args, context, info) => {
  const password = await bcrypt.hash(args.password, 10);
  const author = await context.prisma.mutation.createAuthor(
    {
      data: {
        ...args,
        password
      }
    },
    `{ id }`
  );

  const token = jwt.sign(
    {
      authorId: author.id
    },
    process.env.APP_SECRET
  );

  return {
    token,
    author
  };
};

const login = async (root, args, context, info) => {
  const author = await context.prisma.query.author(
    {
      where: {
        email: args.email
      }
    },
    `{ id password }`
  );

  if (!author) {
    throw new Error(`No author with the email ${args.email}.`);
  }

  const valid = await bcrypt.compare(args.password, author.password);

  if (!valid) {
    throw new Error(`The password is incorrect.`);
  }

  const token = jwt.sign(
    {
      authorId: author.id
    },
    process.env.APP_SECRET
  );

  return {
    token,
    author
  };
};

const createActivity = async (root, args, context, info) => {
  const { tripID, title, locations, price, startDate, endDate, type } = args;
  const trip = await context.prisma.query.trip(
    {
      where: {
        id: tripID
      }
    },
    `{id}`
  );

  if (!trip) {
    throw new Error(`No trip with id ${tripID}`);
  }

  const newActivity = await context.prisma.mutation.createActivity(
    {
      data: {
        title,
        locations,
        price,
        startDate,
        endDate,
        type
      }
    },
    `{ id }`
  );
  context.prisma.mutation.updateTrip({
    data: {
      activities: {
        connect: {
          id: newActivity.id
        }
      }
    },
    where: {
      id: tripID
    }
  });

  return getActivity(newActivity.id, context, info);
};

const createTrip = async (root, args, context, info) => {
  const { authorId, title } = args;
  const author = await context.prisma.query.author(
    {
      where: {
        id: authorId
      }
    },
    `{id}`
  );

  if (!author) {
    throw new Error(`No author with id ${authorId}`);
  }

  const newTrip = await context.prisma.mutation.createTrip(
    {
      data: {
        title
      }
    },
    `{ id }`
  );
  context.prisma.mutation.updateAuthor({
    data: {
      trips: {
        connect: {
          id: newTrip.id
        }
      }
    },
    where: {
      id: authorId
    }
  });

  return getTrip(newTrip.id, context, info);
};

const updateActivity = async (root, args, context, info) => {
  const {
    activityID,
    title,
    locations,
    price,
    startDate,
    endDate,
    type,
    description
  } = args;

  return context.prisma.mutation.updateActivity({
    data: {
      title,
      locations,
      price,
      startDate,
      endDate,
      type,
      description
    },
    where: {
      id: activityID
    }
  }, info)
};

const deleteActivity = async (root, args, context, info) => {
  return context.prisma.mutation.deleteActivity({
    where: {
      id: args.activityID
    }
  }, info)
};

const requestFriendship = async (root, args, context, info) => {
  const { authorId, authorRequestedId } = args;

  context.prisma.mutation.updateAuthor({
    data: {
      friends: {
        connect: {
          id: authorRequestedId
        }
      }
    },
    where: {
      id: authorId
    }
  });

  context.prisma.mutation.updateAuthor({
    data: {
      friendshipRequested: {
        connect: {
          id: authorId
        }
      }
    },
    where: {
      id: authorRequestedId
    }
  });
};

module.exports = {
  signup,
  login,
  createActivity,
  updateActivity,
  deleteActivity,
  createTrip,
  requestFriendship
};
