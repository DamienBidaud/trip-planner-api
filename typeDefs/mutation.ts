import { gql } from "apollo-server";

export default gql`
  type Mutation {
    signup(email: String!, password: String!, username: String!): AuthPayload!
    login(email: String!, password: String!): AuthPayload!
  }
`;
