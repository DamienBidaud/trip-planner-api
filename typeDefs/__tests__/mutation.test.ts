import mutation from "../mutation";

describe("typeDefs", () => {
  describe("mutation", () => {
    it("Should match snapshot", () => {
      expect(mutation).toMatchSnapshot();
    });
  });
});
