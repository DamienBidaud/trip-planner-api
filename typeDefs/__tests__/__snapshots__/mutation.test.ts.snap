// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`typeDefs mutation Should match snapshot 1`] = `
Object {
  "definitions": Array [
    Object {
      "description": undefined,
      "directives": Array [],
      "fields": Array [
        Object {
          "arguments": Array [
            Object {
              "defaultValue": undefined,
              "description": undefined,
              "directives": Array [],
              "kind": "InputValueDefinition",
              "name": Object {
                "kind": "Name",
                "value": "email",
              },
              "type": Object {
                "kind": "NonNullType",
                "type": Object {
                  "kind": "NamedType",
                  "name": Object {
                    "kind": "Name",
                    "value": "String",
                  },
                },
              },
            },
            Object {
              "defaultValue": undefined,
              "description": undefined,
              "directives": Array [],
              "kind": "InputValueDefinition",
              "name": Object {
                "kind": "Name",
                "value": "password",
              },
              "type": Object {
                "kind": "NonNullType",
                "type": Object {
                  "kind": "NamedType",
                  "name": Object {
                    "kind": "Name",
                    "value": "String",
                  },
                },
              },
            },
            Object {
              "defaultValue": undefined,
              "description": undefined,
              "directives": Array [],
              "kind": "InputValueDefinition",
              "name": Object {
                "kind": "Name",
                "value": "username",
              },
              "type": Object {
                "kind": "NonNullType",
                "type": Object {
                  "kind": "NamedType",
                  "name": Object {
                    "kind": "Name",
                    "value": "String",
                  },
                },
              },
            },
          ],
          "description": undefined,
          "directives": Array [],
          "kind": "FieldDefinition",
          "name": Object {
            "kind": "Name",
            "value": "signup",
          },
          "type": Object {
            "kind": "NonNullType",
            "type": Object {
              "kind": "NamedType",
              "name": Object {
                "kind": "Name",
                "value": "AuthPayload",
              },
            },
          },
        },
        Object {
          "arguments": Array [
            Object {
              "defaultValue": undefined,
              "description": undefined,
              "directives": Array [],
              "kind": "InputValueDefinition",
              "name": Object {
                "kind": "Name",
                "value": "email",
              },
              "type": Object {
                "kind": "NonNullType",
                "type": Object {
                  "kind": "NamedType",
                  "name": Object {
                    "kind": "Name",
                    "value": "String",
                  },
                },
              },
            },
            Object {
              "defaultValue": undefined,
              "description": undefined,
              "directives": Array [],
              "kind": "InputValueDefinition",
              "name": Object {
                "kind": "Name",
                "value": "password",
              },
              "type": Object {
                "kind": "NonNullType",
                "type": Object {
                  "kind": "NamedType",
                  "name": Object {
                    "kind": "Name",
                    "value": "String",
                  },
                },
              },
            },
          ],
          "description": undefined,
          "directives": Array [],
          "kind": "FieldDefinition",
          "name": Object {
            "kind": "Name",
            "value": "login",
          },
          "type": Object {
            "kind": "NonNullType",
            "type": Object {
              "kind": "NamedType",
              "name": Object {
                "kind": "Name",
                "value": "AuthPayload",
              },
            },
          },
        },
      ],
      "interfaces": Array [],
      "kind": "ObjectTypeDefinition",
      "name": Object {
        "kind": "Name",
        "value": "Mutation",
      },
    },
  ],
  "kind": "Document",
  "loc": Object {
    "end": 161,
    "start": 0,
  },
}
`;
